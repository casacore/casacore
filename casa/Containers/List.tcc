//# List.cc: Singly linked list classes
//# Copyright (C) 1993,1994,1995,1997,1998,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/Containers/List.h>
#include <casa/Containers/IterError.h>

namespace casa { //#Begin casa namespace

template<class t> uInt ListNotice<t>::type() const  {
  return Register(this);
}

template<class t> int ListNotice<t>::operator==(const Notice &op) const {
  if (type() != op.type()) {
    return 0;
  } else {
    const ListNotice<t> &opD = (const ListNotice<t> &) op;
    return (mod == opD.mod && ocur == opD.ocur && oprev == opD.oprev && 
	    off == opD.off && otherOff == opD.otherOff );
  }
}

template<class t> void List<t>::added(Link<t> *prev, Link<t> *cur) {
  length++;
  if (head == 0 || prev == 0) {
    head = cur;
  }
  if (tail == 0 || tail == prev) {
    tail = cur;
  }
}

template<class t> void List<t>::removed(Link<t> *oldl, Link<t> *prev, Link<t> *cur) { 
  length--;
  if (oldl == head) head = cur;    
  if (oldl == tail) tail = prev;
}

template<class t> List<t>::List(const List<t> &other) : head(0), tail(0), length(0) {
    ConstListIter<t> iter(other);
    Link<t> *cur = 0;
    iter.toStart();
    if (! iter.atEnd()) {
	head = cur = new Link<t>(iter.getRight());
	iter++;
    }
    for (; !iter.atEnd(); iter++ )
	cur = new Link<t>(iter.getRight(),cur);
    tail = cur;
    length = other.len();
}

template<class t> List<t>::List(const List<t> *other) : head(0), tail(0), length(0) {
    ConstListIter<t> iter(other);
    Link<t> *cur = 0;
    iter.toStart();
    if (! iter.atEnd()) {
	head = cur = new Link<t>(iter.getRight());
	iter++;
    }
    for (; !iter.atEnd(); iter++ )
	cur = new Link<t>(iter.getRight(),cur);
    tail = cur;
    length = other->len();
}

template<class t> List<t> &List<t>::operator=(const List<t> &other) {
    if ( this == &other )
	return *this;
    ConstListIter<t> iter(other);
    Link<t> *cur = 0;
    if ( head ) delete head;
    head = tail = 0;
    length = 0;
    if (! iter.atEnd()) {
	head = cur = new Link<t>(iter.getRight());
	iter++;
    }
    for (; !iter.atEnd(); iter++ )
	cur = new Link<t>(iter.getRight(),cur);
    tail = cur;
    length = other.len();
    return *this;
}
    
template<class t> List<t> &List<t>::operator=(const List<t> *other) {
    if ( this == other )
	return *this;
    ConstListIter<t> iter(other);
    Link<t> *cur = 0;
    if ( head ) delete head;
    head = tail = 0;
    length = 0;
    if (! iter.atEnd()) {
	head = cur = new Link<t>(iter.getRight());
	iter++;
    }
    for (; !iter.atEnd(); iter++ )
	cur = new Link<t>(iter.getRight(),cur);
    tail = cur;
    length = other->len();
    return *this;
}

    
template<class t> List<t>::~List() {
  if (head) {
    ListNotice<t> state;
    notify(state);
    delete head;
  }
}    

template<class t>
ConstListIter<t>::ConstListIter(const List<t> *st) :  
                                     NoticeTarget((NoticeSource *)st),
                                     cur((*st).head), prev(0),
                                     curPos(0), container_((List<t> *)st) {}

template<class t>
ConstListIter<t>::ConstListIter(const ConstListIter<t> *other) : NoticeTarget((NoticeTarget *)other) {
  if (other) {
    cur = (*other).cur;
    prev = (*other).prev;
    curPos = (*other).curPos;
    container_ = (*other).container_;
  } else throw_list_init_error();
}

template<class t>
ConstListIter<t>::~ConstListIter() {}

template<class t>
ConstListIter<t> &ConstListIter<t>::operator=(const List<t> &other) {
  cur = other.head;
  prev = 0;
  curPos = 0;
  container_ = (List<t> *) &other;
  attach((List<t> &)other);
  return *this;
}

template<class t>
ConstListIter<t> &ConstListIter<t>::operator=(const List<t> *other) {
  if (other) {
    cur = (*other).head;
    prev = 0;
    curPos = 0;
    container_ = (List<t> *) other;
    attach((List<t> *)other);
  }
  return *this;
}

template<class t>
ConstListIter<t> &ConstListIter<t>::operator=(const ConstListIter<t> &other) {
  if (other.isValid()) {
    cur = other.cur;
    prev = other.prev;
    curPos = other.curPos;
    container_ = other.container_;
    link(other);
  }
  return *this;
}

template<class t>
ConstListIter<t> &ConstListIter<t>::operator=(const ConstListIter<t> *other) {
  if (other && (*other).isValid()) {
    cur = (*other).cur;
    prev = (*other).prev;
    curPos = (*other).curPos;
    container_ = (*other).container_;
    link(other);
  }
  return *this;
}


template<class t> void ConstListIter<t>::notify(const Notice &note) {
  if (Register(static_cast<ListNotice<t> *>(0)) == note.type()) {
    const ListNotice<t> &opD = (const ListNotice<t> &) note;
    if ( opD.mod != ListNotice<t>::DELETE ) {
      if (cur == opD.ocur && prev == opD.oprev) {
	cur = opD.ncur;
	prev = opD.nprev;
      }
      if ( Int(curPos) > opD.off ) {
	switch ( opD.mod ) {
	case ListNotice<t>::ADD:
		++curPos; break;
	case ListNotice<t>::REMOVE:
		--curPos; break;
	case ListNotice<t>::SWAP:
		curPos += opD.otherOff - opD.off; break;
	default:
	        break;
	}
      }
    } else {
      invalidate();
      container_ = 0;
    }
  }
}

template<class t> uInt ConstListIter<t>::pos(uInt where) {
  uInt i;

  AlwaysAssert(isValid(),InvalidIterError);

  if (where > (*container_).length) 
    throw(IterBoundaryError("Iterator advanced past the end of the list."));
  if (where == curPos) 
    return(curPos);
  if (where > curPos ) {
    uInt savedPos = curPos;
    for (i = 0; i < where - savedPos && atEnd() != True; i++) 
      operator++();
  } else {
    toStart();
    for (i = 0; i < where; i++) operator++();
  }
  return(curPos);
}

template<class t>
ListIter<t>::ListIter(List<t> &st) : ConstListIter<t>(st), own(False){}

template<class t>
ListIter<t>::ListIter(const ListIter<t> &other) : ConstListIter<t>((const ConstListIter<t> &) other), own(False){}

template<class t>
ListIter<t>::~ListIter() {
  if (own && this->container_)
    delete this->container_;
}

template<class t>
ListIter<t> &ListIter<t>::assign(List<t> *other,Bool OWN) {
  if (other && own && this->container_) {
    delete this->container_;
  }
  own = OWN;
  ConstListIter<t>::operator=(other);
  return *this;
}

template<class t>
Link<t> *ListIter<t>::newLink(t &e, Link<t> *p, Link<t> *n) {
  return new Link<t>(e,p,n);
}


template<class t>
ListIter<t> &ListIter<t>::operator=(List<t> &other) {
  if (own && this->container_)
    delete this->container_;
  ConstListIter<t>::operator=(other);
  own = False;
  return *this;
}

template<class t>
ListIter<t> &ListIter<t>::operator=(List<t> *other) {
  if (other && own && this->container_)
    delete this->container_;
  ConstListIter<t>::operator=(other);
  own = False;
  return *this;
}

template<class t>
ListIter<t> &ListIter<t>::operator=(const ListIter<t> &other) {
  if (other.isValid() && own && this->container_)
    delete this->container_;
  ConstListIter<t>::operator=(other);
  own = False;
  return *this;
}

template<class t>
ListIter<t> &ListIter<t>::operator=(const ListIter<t> *other) {
  if (other && (*other).isValid() && own && this->container_)
    delete this->container_;
  ConstListIter<t>::operator=(other);
  own = False;
  return *this;
}

template<class t> void ListIter<t>::removeRight() {

  AlwaysAssert(this->isValid(),InvalidIterError);

  if (this->cur != 0) {
    Link<t> *c = this->cur;
    Link<t> *p = this->prev;
    this->cur = (*this->cur).unlink(this->prev);  // Link<t>::unlink: 
                                        //         * sets links for deletion
                                        //         * returns next
    // Allow container to update
    (*this->container_).removed(c,this->prev,this->cur);
    ListNotice<t> state(ListNotice<t>::REMOVE,c,p,this->cur,this->prev,
			this->curPos);
    (*this->container_).notify(state);
    delete c;
  } else throw_list_end_error();
}

template<class t> void ListIter<t>::swapRight(ListIter<t> &swapee){
  if (this->container_ == swapee.container_) throw_list_swapright_same_error();
  Link<t> *mc = this->cur;
  Link<t> *mp = this->prev;
  Link<t> *sc = swapee.cur;
  Link<t> *sp = swapee.prev;
  Link<t> *tmpTail = (*this->container_).tail;
  if (this->prev != 0) {
    (*this->prev).next() = swapee.cur;
    if (swapee.cur) {
      (*swapee.cur).prev() = this->prev;
      (*this->container_).tail = (*swapee.container_).tail;
    } else {
      (*this->container_).tail = this->prev;
    }
  } else {
    (*this->container_).head = swapee.cur;
    if (swapee.cur) {
      (*swapee.cur).prev() = 0;
      (*this->container_).tail = (*swapee.container_).tail;
    } else {
      (*this->container_).tail = this->prev;
    }
  }
  Link<t> *tmp = this->cur;
  this->cur = swapee.cur;
  if (swapee.prev != 0) {
    (*swapee.prev).next() = tmp;
    if (tmp) {
      (*tmp).prev() = swapee.prev;
      (*swapee.container_).tail = tmpTail;
    } else {
      (*swapee.container_).tail = swapee.prev;
    }
  } else {
    (*swapee.container_).head = tmp;
    if (tmp) {
      (*tmp).prev() = 0;
      (*swapee.container_).tail = tmpTail;
    } else {
      (*swapee.container_).tail = swapee.prev;
    }
  }
  swapee.cur = tmp;

  // Adjust lengths
  uInt tmpLen = (*this->container_).length;
  (*this->container_).length = (this->curPos - 1) + swapee.len() -
                               (swapee.pos() - 1);
  (*swapee.container_).length = (swapee.pos() - 1) + tmpLen -
                                (this->curPos - 1);

  ListNotice<t> mstate(ListNotice<t>::SWAP,mc,mp,this->cur,this->prev,
		       this->curPos,swapee.curPos);
  (*this->container_).notify(mstate);

  ListNotice<t> sstate(ListNotice<t>::SWAP,sc,sp,swapee.cur,swapee.prev,
		       swapee.curPos,this->curPos);
  (*swapee.container_).notify(sstate);
}

//#
//# These functions are protected and ONLY throw an exception
//# to prevent improper initialization of a constant OrderedMapIter
//#
template<class t>
ConstListIter<t> &ListIter<t>::operator=(const List<t> &) {
  throw_list_init_error();
  return *this;}
template<class t>
ConstListIter<t> &ListIter<t>::operator=(const List<t> *) {
  throw_list_init_error();
  return *this;}
template<class t>
ConstListIter<t> &ListIter<t>::operator=(const ConstListIter<t> &) {
  throw_list_init_error();
  return *this;}
template<class t>
ConstListIter<t> &ListIter<t>::operator=(const ConstListIter<t> *) {
  throw_list_init_error();
  return *this;}

} //#End casa namespace
