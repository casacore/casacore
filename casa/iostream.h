//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
