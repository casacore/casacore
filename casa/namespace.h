#ifndef NAMESPACE_H
#define NAMESPACE_H 
using namespace casa;
#endif
