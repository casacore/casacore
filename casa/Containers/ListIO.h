//# ListIO.h: Singly linked list IO
//# Copyright (C) 1993,1994,1995,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_LISTIO_H
#define CASA_LISTIO_H

#include <casacore/casa/Containers/List.h>


namespace casa { //# NAMESPACE CASA - BEGIN

// <summary>
// Input/output operators for Lists.
// </summary>

// <use visibility=export>

// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=List>List</linkto>
//   <li> ostream
// </prerequisite>

// <synopsis> 
// These global functions provide easy input and output of List objects.
// </synopsis>
//
// <group name="List IO">

// These functions are used to write <src>List</src>s to an ostream.
// <group>
template<class t> ostream &operator<<(ostream &, const List<t> &);
template<class t> ostream &operator<<(ostream &, const ConstListIter<t> &);
// </group>
// </group>


} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casacore/casa/Containers/ListIO.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
