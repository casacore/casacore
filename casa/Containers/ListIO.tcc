//# ListIO.cc: Doubly linked list IO
//# Copyright (C) 1993,1994,1995,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Containers/ListIO.h>


namespace casa { //# NAMESPACE CASA - BEGIN

//
//  Outputs a doubly linked list to the specified ostream stream.
//
template<class t> ostream &operator<<(ostream &ios, const List<t> &list) {
  ConstListIter<t> listp = list;

  ios << "len=" << list.len();
  listp.toStart();
  while (listp.atEnd() == False) {
    ios << (char) ios.fill() << listp.getRight();
    listp++;
  }
  return(ios);
}

//
//  Outputs a doubly linked list iterator to the specified ostream stream.
//
template<class t> ostream &operator<<(ostream &ios, const ConstListIter<t> &list) {
  ConstListIter<t> listp = list;

  ios << "len=" << list.len() << (char) ios.fill() << "pos=" << list.pos();
  listp.toStart();
  while (listp.atEnd() == False) {
    ios << (char) ios.fill() << listp.getRight();
    listp++;
  }
  return(ios);
}

} //# NAMESPACE CASA - END

