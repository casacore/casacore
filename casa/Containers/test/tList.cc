//# tList.cc: This program tests the Dlist class
//# Copyright (C) 1993,1994,1995,1998,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes

#include <casa/Containers/List.h>
#include <casa/Containers/ListIO.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
static void show(const ConstListIter<int> &);
static void dump(const ConstListIter<int> *);
int main() {

  List<int> one;
  ListIter<int> onePa(one);
  ListIter<int> onePb(one);

  onePa.addRight(3);
  onePa.addRight(72);
  onePa.addRight(16);
  onePa.addRight(7);
  onePa.addRight(31);

  cout << "-- -- -- -- -- (1) -- -- -- -- --" << endl;

  cout << "<onePa> - ";
  while(onePa.atEnd() != True) {
    cout << onePa.getRight() << ".";
    onePa++;
  }
  cout << endl;

  cout << "<onePa> - ";
  onePa--;
  while(onePa.atStart() != True) {
    cout << onePa.getRight() << ".";
    onePa--;
  }
  cout << onePa.getRight() << ".";
  cout << endl;

  cout << "-- -- -- -- -- (2) -- -- -- -- --" << endl;

  onePa.pos(3);
  cout << onePa.getRight() << " & ";
  onePb.pos(3);
  cout << onePb.getRight() << endl;

  onePa.addRight(103);
  
  cout << onePa.getRight() << " & ";
  cout << onePb.getRight() << endl;

  cout << "<b:onePb> - ";
  onePb.toEnd();
  if (onePb.atStart() != True) do {
    onePb--;
    cout << onePb.getRight() << ".";
  } while (onePb.atStart() != True);
  cout << endl;
  cout << "<f:onePb> - ";
  onePb.toStart();
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;

  cout << "-- -- -- -- -- (3) -- -- -- -- --" << endl;

  cout << onePa.getRight() << " <30> ";
  onePa.step(30);
  cout << onePa.getRight() << " <30> ";
  onePa.step(30);
  cout << onePa.getRight() << " <-25> ";
  onePa.step(-25);
  cout << onePa.getRight() << " <13> ";
  onePa.step(13);
  cout << onePa.getRight() << endl;

  cout << "-- -- -- -- -- (4) -- -- -- -- --" << endl;

  onePa.pos(4);
  onePb.pos(4);
  onePb.removeRight();
  cout << onePa.getRight() << " & ";
  cout << onePb.getRight() << endl;

  cout << "<b:onePb> - ";
  onePb.toEnd();
  if (onePb.atStart() != True) do {
    onePb--;
    cout << onePb.getRight() << ".";
  } while(onePb.atStart() != True);
  cout << endl;
  cout << "<f:onePb> - ";
  onePb.toStart();
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;

  onePa--;
  onePb--;
  --onePa;
  --onePb;
  onePa.removeRight();
  cout << onePa.getRight() << " & ";
  cout << onePb.getRight() << endl;

  cout << "<b:onePb> - ";
  onePb.toEnd();
  if (onePb.atStart() != True) do {
    onePb--;
    cout << onePb.getRight() << ".";
  } while(onePb.atStart() != True);
  cout << endl;
  cout << "<f:onePb> - ";
  onePb.toStart();
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;

  cout << "-- -- -- -- -- (5) -- -- -- -- --" << endl;

  --onePa;

  cout << "<onePc> - ";
  ListIter<int> onePc(&onePa);
  while(onePc.atEnd() != True) {
    cout << onePc.getRight() << ".";
    onePc++;
  }
  cout << endl;
  List<int> two;
  onePc = two;

  onePc.addRight(604);
  onePc.addRight(671);
  onePc.addRight(601);
  onePc.addRight(685);
  onePc.addRight(690);
  onePc.addRight(613);
  onePc.addRight(654);
  onePc.addRight(623);
  onePc.addRight(611);

  cout << "<onePc> - ";
  while(onePc.atEnd() != True) {
    cout << onePc.getRight() << ".";
    onePc++;
  }
  cout << endl;

  onePc.pos(4);
  onePa.pos(2);
  onePb.pos(2);

  onePc.swapRight(onePa);

  cout << "<onePb> - ";
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;

  cout << "<b:onePc> - ";
  onePc.toEnd();
  if (onePc.atStart() != True) do {
    onePc--;
    cout << onePc.getRight() << ".";
  } while(onePc.atStart() != True);
  cout << endl;
  cout << "<f:onePc> - ";
  onePc.toStart();
  while(onePc.atEnd() != True) {
    cout << onePc.getRight() << ".";
    onePc++;
  }
  cout << endl;

  cout << "<b:onePa> - ";
  onePa.toEnd();
  if (onePa.atStart() != True) do {
    onePa--;
    cout << onePa.getRight() << ".";
  } while(onePa.atStart() != True);
  cout << endl;
  cout << "<f:onePa> - ";
  onePa.toStart();
  while(onePa.atEnd() != True) {
    cout << onePa.getRight() << ".";
    onePa++;
  }
  cout << endl;

  cout << "-- -- -- -- -- (6) -- -- -- -- --" << endl;

  List<int> three;
  onePb = three;

  onePc.pos(4);
  onePb.swapRight(onePc);

  cout << "<b:onePb> - ";
  onePb.toEnd();
  if (onePb.atStart() != True)  do {
    onePb--;
    cout << onePb.getRight() << ".";
  } while(onePb.atStart() != True);
  cout << endl;
  cout << "<f:onePb> - ";
  onePb.toStart();
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;

  cout << "<b:onePa> - ";
  onePa.toEnd();
  if (onePa.atStart() != True) do {
    onePa--;
    cout << onePa.getRight() << ".";
  } while(onePa.atStart() != True);
  cout << endl;
  cout << "<f:onePa> - ";
  onePa.toStart();
  while(onePa.atEnd() != True) {
    cout << onePa.getRight() << ".";
    onePa++;
  }
  cout << endl;
  onePa.toStart();

  cout << "<b:onePc> - ";
  onePc.toEnd();
  if (onePc.atStart() != True) do {
    onePc--;
    cout << onePc.getRight() << ".";
  } while(onePc.atStart() != True);
  cout << endl;
  cout << "<f:onePc> - ";
  onePc.toStart();
  while(onePc.atEnd() != True) {
    cout << onePc.getRight() << ".";
    onePc++;
  }
  cout << endl;
  onePc.toStart();

  onePb.swapRight(onePa);
  for(;onePb.atEnd() != True;onePb++) {}
  onePb.swapRight(onePc);

  cout << "<b:onePa> - ";
  onePa.toEnd();
  if (onePa.atStart() != True) do {
    onePa--;
    cout << onePa.getRight() << ".";
  } while(onePa.atStart() != True);
  cout << endl;
  cout << "<f:onePa> - ";
  onePa.toStart();
  while(onePa.atEnd() != True) {
    cout << onePa.getRight() << ".";
    onePa++;
  }
  cout << endl;
  onePa.toStart();

  cout << "<b:onePb> - ";
  onePb.toEnd();
  if (onePb.atStart() != True) do {
    onePb--;
    cout << onePb.getRight() << ".";
  } while(onePb.atStart() != True);
  cout << endl;
  cout << "<f:onePb> - ";
  onePb.toStart();
  while(onePb.atEnd() != True) {
    cout << onePb.getRight() << ".";
    onePb++;
  }
  cout << endl;
  onePb.toStart();

  cout << "<b:onePc> - ";
  onePc.toEnd();
  if (onePc.atStart() != True) do {
    onePc--;
    cout << onePc.getRight() << ".";
  } while(onePc.atStart() != True);
  cout << endl;
  cout << "<f:onePc> - ";
  onePc.toStart();
  while(onePc.atEnd() != True) {
    cout << onePc.getRight() << ".";
    onePc++;
  }
  cout << endl;
  onePc.toStart();

  List<int> *four = new List<int>;
  ListIter<int> onePd (four);

  onePd.addRight(23);
  onePd.addRight(24);
  onePd.addRight(29);
  onePd.addRight(20);
  onePd.addRight(22);
  onePd.addRight(28);

  try {
    onePd.toStart();
    while (1) {
      cout << onePd.getRight() << " ";
      onePd++;
    }
  } catch (IterError xx) {
    cout << endl << "IterError: " << xx.getMesg() << endl;
  } 

  try {
    onePd--;
    while (1) {
      cout << onePd.getRight() << " ";
      onePd--;
    }
  } catch (IterError xx) {
    cout << endl << "IterError: " << xx.getMesg() << endl;
  } 

  ConstListIter<int> t5,t6;
  ListIter<int> t7,t8;
  t5 = two;
  t6 = four;
  t7 = two;
  t8 = four;
  show(t5);
  show(t6);

  try {
    onePd.toStart();
    while (!onePd.atEnd()) {
      cout << onePd.getRight() << " ";
      onePd++;
    }
    cout << endl;
    onePd.toStart();
    delete four;
    while (!onePd.atEnd()) {
      cout << onePd.getRight() << " ";
      onePd++;
    }
  } catch (IterError xx) {
    cout << endl << ">>> Instance-specific assertion error message:" << endl
         << "#X# IterError: " << xx.getMesg() << endl
         << "<<< End of assertion error message." << endl;
  } 


  ConstListIter<int> t1,t2;
  ListIter<int> t3,t4;
  
  t1 = onePa;
  cout << "-- -- -- -- -- (cout) -- -- -- -- --" << endl;
  cout << t1 << endl;
  t2 = &onePb;
  show(t2);
  t3 = onePc;
  dump(&t3);
  try {
    cout << ">>> Instance-specific assertion error message:" << endl;
    t4 = &onePd;
    show(t4);
  } catch (IterError xx) {
    cout << endl
         << "#X# IterError: " << xx.getMesg() << endl
         << "<<< End of assertion error message." << endl;
  } 

  ListIter<int> oth;
  oth.assign(new List<int>(),True);
  oth.addRight(3);
  oth.addRight(4);
  oth.addRight(5);
  oth.addRight(6);
  show(oth);

  List<int> oth1(oth.container());
  ListIter<int> last_a(&oth1);
  show( last_a );
		  

  return 0;
}

static void show(const ConstListIter<int> &i) {
    cout << "-- -- -- -- -- (show) -- -- -- -- --" << endl;
    cout << i << endl;
}

static void dump(const ConstListIter<int> *i) {
    ConstListIter<int> list(i);
    cout << "-- -- -- -- -- (dump) -- -- -- -- --" << endl;
    cout << "len=" << list.len() << (char) cout.fill() <<  "pos=" << list.pos();
    list.toStart();
    while (list.atEnd() == False) {
	cout << (char) cout.fill() << list.getRight();
	list++;
    }
    cout << endl;
}
