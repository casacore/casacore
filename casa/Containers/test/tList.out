-- -- -- -- -- (1) -- -- -- -- --
<onePa> - 31.7.16.72.3.
<onePa> - 3.72.16.7.31.
-- -- -- -- -- (2) -- -- -- -- --
72 & 72
103 & 103
<b:onePb> - 3.72.103.16.7.31.
<f:onePb> - 31.7.16.103.72.3.
-- -- -- -- -- (3) -- -- -- -- --
103 <30> 3 <30> 31 <-25> 103 <13> 16
-- -- -- -- -- (4) -- -- -- -- --
3 & 3
<b:onePb> - 3.103.16.7.31.
<f:onePb> - 31.7.16.103.3.
103 & 103
<b:onePb> - 3.103.7.31.
<f:onePb> - 31.7.103.3.
-- -- -- -- -- (5) -- -- -- -- --
<onePc> - 7.103.3.
<onePc> - 611.623.654.613.690.685.601.671.604.
<onePb> - 690.685.601.671.604.
<b:onePc> - 3.103.613.654.623.611.
<f:onePc> - 611.623.654.613.103.3.
<b:onePa> - 604.671.601.685.690.7.31.
<f:onePa> - 31.7.690.685.601.671.604.
-- -- -- -- -- (6) -- -- -- -- --
<b:onePb> - 3.103.
<f:onePb> - 103.3.
<b:onePa> - 604.671.601.685.690.7.31.
<f:onePa> - 31.7.690.685.601.671.604.
<b:onePc> - 613.654.623.611.
<f:onePc> - 611.623.654.613.
<b:onePa> - 
<f:onePa> - 
<b:onePb> - 613.654.623.611.604.671.601.685.690.7.31.3.103.
<f:onePb> - 103.3.31.7.690.685.601.671.604.611.623.654.613.
<b:onePc> - 
<f:onePc> - 
28 22 20 29 24 23 
IterError: List operation performed with cursor past the end of the list.
23 24 29 20 22 28 
IterError: Iterator backed past the beginning of the list.
-- -- -- -- -- (show) -- -- -- -- --
len=0 pos=0
-- -- -- -- -- (show) -- -- -- -- --
len=6 pos=0 28 22 20 29 24 23
28 22 20 29 24 23 

>>> Instance-specific assertion error message:
#X# IterError: (/aips++/code/include/aips/Containers/List.h : 378) Failed AlwaysAssertExit isValid()
<<< End of assertion error message.
-- -- -- -- -- (cout) -- -- -- -- --
len=0 pos=0
-- -- -- -- -- (show) -- -- -- -- --
len=13 pos=0 103 3 31 7 690 685 601 671 604 611 623 654 613
-- -- -- -- -- (dump) -- -- -- -- --
len=0 pos=0
>>> Instance-specific assertion error message:
-- -- -- -- -- (show) -- -- -- -- --
len=
#X# IterError: (/aips++/code/include/aips/Containers/List.h : 443) Failed AlwaysAssertExit isValid()
<<< End of assertion error message.
-- -- -- -- -- (show) -- -- -- -- --
len=4 pos=0 6 5 4 3
-- -- -- -- -- (show) -- -- -- -- --
len=4 pos=0 6 5 4 3
